File: app/src/at-phuongle/java/com/asiantech/intern20summer1/intro/LandOfLogic.java
package com.asiantech.intern20summer1.intro;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class LandOfLogic {
    /**
     * Write a function that returns the longest word from the given string.
     */
    String longestWord(String text) {
        String[] strings = text.split("[^a-zA-Z]+");
        String longestWord = "";

        for (int i = 0; i < strings.length; i++) {
            if (strings[i].length() > longestWord.length()) {
                longestWord = strings[i];
            }
        }

        return longestWord;
    }

    /**
     * Write a function that checks if the given string is a correct time representation of the 24-hour clock.
     */
    boolean validTime(String time) {
        String[] strings = time.split("[:]");
        if (Integer.parseInt(strings[0]) < 0 || Integer.parseInt(strings[0]) > 23) {
            return false;
        }
        if (Integer.parseInt(strings[1]) < 0 || Integer.parseInt(strings[1]) > 59) {
            return false;
        }

        return true;
    }

    /**
     * Write a function that returns the sum of numbers that appear in the given inputString.
     */
    int sumUpNumbers(String inputString) {
        int sum = 0;
        String[] strings = inputString.split("[^0-9]+");

        for (int i = 0; i < strings.length; i++) {
            if (!strings[i].isEmpty()) {
                sum += Integer.parseInt(strings[i]);
            }
        }

        return sum;
    }

    /**
     * Write a function that returns the decrypted message.
     */
    String messageFromBinaryCode(String code) {
        String result = "";
        String[] strings = code.split("(?<=\\G.{8})");

        for (int i = 0; i < strings.length; i++) {
            result += (char) Integer.parseInt(strings[i], 2);
        }

        return result;
    }

    /**
     * Write a function that checks if the given array is validate.
     */
    boolean validate(int[] check) {
        int i = 0;
        Arrays.sort(check);
        for (int number : check) {
            if (number != ++i)
                return false;
        }
        return true;
    }

    /**
     * Write a function that checks if the given grid of numbers represents a correct solution to Sudoku.
     */
    boolean sudoku(int[][] grid) {
        for (int i = 0; i < 9; i++) {

            int[] row = new int[9];
            int[] square = new int[9];
            int[] column = grid[i].clone();

            for (int j = 0; j < 9; j ++) {
                row[j] = grid[j][i];
                square[j] = grid[(i / 3) * 3 + j / 3][i * 3 % 9 + j % 3];
            }
            if (!(validate(column) && validate(row) && validate(square)))
                return false;
        }
        return true;
    }

    /**
     * Write a function that returns an array of names that will be given to the files.
     */
    String[] fileNaming(String[] names) {
        List<String> result = new ArrayList<String>();
        for (int i = 0; i < names.length; i++) {
            int count = 0;
            String s = names[i];
            while (result.contains(s)) {
                count++;
                s = names[i] + "(" + count + ")";
            }
            result.add(s);
        }

        return result.toArray(new String[result.size()]);
    }

    /**
     * Write a function that returns the smallest positive integer the product of whose digits is equal to product.
     */
    int digitsProduct(int product) {
        if (product == 0) {
            return 10;
        }
        if (product == 1) {
            return 1;
        }
        String digits = "";
        for (int divisor = 9; divisor > 1; divisor--) {
            while (product % divisor == 0) {
                product /= divisor;
                digits = (divisor) + digits;
            }
        }

        if (product > 1) return -1;

        return Integer.parseInt(digits);
    }
}

File: app/src/at-phuongle/java/com/asiantech/intern20summer1/kotlin/LandOfLogic.kt
package com.asiantech.intern20summer1.kotlin

object LandOfLogic {
    private const val NINE = 9
    private const val VALID_HOUR = 23
    private const val VALID_MINUTE = 59

    fun longestWord(text: String): String {
        var regex: String = "[^a-zA-Z]+"
        var strings: Array<String> = text.split(regex.toRegex()).toTypedArray()
        var longestWord: String = ""

        for (i in strings.indices) {
            if (strings[i].length > longestWord.length) {
                longestWord = strings[i]
            }
        }

        return longestWord
    }

    fun validTime(time: String): Boolean {
        var check: Boolean = true
        var regex: String = "[:]"
        var strings: Array<String> = time.split(regex.toRegex()).toTypedArray()
        if (strings[0].toInt() < 0 || strings[0].toInt() > VALID_HOUR) {
            check = false
        }
        if (strings[1].toInt() < 0 || strings[1].toInt() > VALID_MINUTE) {
            check = false
        }

        return check
    }

    fun sumUpNumbers(inputString: String): Int {
        var regex: String = "[^0-9]+"
        var sum: Int = 0
        var strings: Array<String> = inputString.split(regex.toRegex()).toTypedArray()

        for (i in strings.indices) {
            if (strings[i].isNotEmpty()) {
                sum += strings[i].toInt()
            }
        }

        return sum
    }

    fun digitsProduct(product: Int): Int {
        var p: Int = product
        var digits: String = ""

        if (p == 0) {
            digits = "10"
        }

        if (p == 1) {
            digits = "1"
        }


        for (divisor in NINE downTo 2) {
            while (p % divisor == 0) {
                p /= divisor
                digits = divisor.toString() + digits
            }
        }

        if (p > 1) {
            digits = "-1"
        }

        return digits.toInt()
    }

    fun fileNaming(names: Array<String>): Array<String> {
        var result: MutableList<String> = mutableListOf()
        for (i in names.indices) {
            var count: Int = 0
            var s: String = names[i]
            while (result.contains(s)) {
                count++
                s = names[i] + "(" + count + ")"
            }
            result.add(s)
        }

        return result.toTypedArray()
    }

    fun messageFromBinaryCode(code: String): String {
        var regex: String = "(?<=\\G.{8})"
        var result: String = ""
        var strings: Array<String> = code.split(regex.toRegex()).toTypedArray()

        for (i in 0 until strings.size - 1) {
            result += strings[i].toInt(2).toChar()
        }

        return result
    }
}
